tsl/src/compression/compression.c

```diff
@@ -1512,3 +1516,155 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
+	}
+	in_column_offsets =
+		compress_chunk_populate_keys(RelationGetRelid(in_rel), ccinfo, cclen, &n_keys, &keys);
+	row_compressor_init(&cr->row_compressor,
+						in_desc,
+						out_rel,
+						cclen,
+						ccinfo,
+						in_column_offsets,
+						out_desc->natts);
+	return cr;
+}
+
+/* create a single row compressed tuple from data in slot*/
+TupleTableSlot *
+compress_row_exec(CompressSingleRowState *cr, TupleTableSlot *slot)
+{
+	TupleTableSlot *compress_slot;
+	slot_getallattrs(slot);
+
+	cr->row_compressor.rows_compressed_into_current_value = 0;
+	row_compressor_update_group(&cr->row_compressor, slot);
+	row_compressor_append_row(&cr->row_compressor, slot);
+	compress_slot = compress_singlerow(cr, slot);
+	return compress_slot;
+}
+
+static TupleTableSlot *
+compress_singlerow(CompressSingleRowState *cr, TupleTableSlot *in_slot)
+{
+	Datum *invalues, *out_values;
+	bool *out_isnull;
+	TupleTableSlot *out_slot = cr->out_slot;
+	RowCompressor *row_compressor = &cr->row_compressor;
+
+	slot_getallattrs(in_slot);
+	ExecClearTuple(out_slot);
+
+	invalues = in_slot->tts_values;
+	out_values = out_slot->tts_values;
+	out_isnull = out_slot->tts_isnull;
+
+	/* Possible optimization:
+	 * Can we do a pass through compression without a full copy?
+	 * full copy needed for multiple values. But we are dealing only with a single value,
+	 * so just need the result of transformation after passing it through the compressor function
+	 * This probably needs a bit of rewrte of the compression algorithm code
+	 */
+	for (int col = 0; col < row_compressor->n_input_columns; col++)
+	{
+        PerColumn *column = &row_compressor->per_column[col];
+		Compressor *compressor = row_compressor->per_column[col].compressor;
+		int in_attrno = col;
+
+		int16 out_attrno = row_compressor->uncompressed_col_to_compressed_col[col];
+		/* if there is no compressor, this must be a segmenter */
+        if (compressor != NULL)
+        {
+            void *compressed_data;
+            compressed_data = compressor->finish(compressor);
+            out_isnull[out_attrno] = (compressed_data == NULL);
+            if (compressed_data)
+                out_values[out_attrno] = PointerGetDatum(compressed_data);
+            if (column->min_max_metadata_builder != NULL)
+            {
+                // should this be copied directly from inslot???
+                out_isnull[column->min_metadata_attr_offset] = false;
+                out_isnull[column->max_metadata_attr_offset] = false;
+                out_values[column->min_metadata_attr_offset] = invalues[in_attrno];
+                out_values[column->max_metadata_attr_offset] = invalues[in_attrno];
+            }
+            else
+            {
+                out_isnull[column->min_metadata_attr_offset] = true;
+                out_isnull[column->max_metadata_attr_offset] = true;
+            }
+        }
+        /* if there is no compressor, this must be a segmenter */
+        else if (column->segment_info != NULL)
```

Comment:
From the comment, it looks like one of these will be true at any time. If that is the case, then maybe you should assert this early. Otherwise there is a risk that, because of a bug, both these conditions are false and we get a random value in `out_isnull` and `out_values`, which can lead to hard to find bugs.

Reply:
Fixed this to throw an error if we ever encounter a bad state.